Fix bug for Parameter with a getter but no setter

larrybradley opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

Currently if a `Parameter` is defined without a `setter`, the `_internal_value` never gets set.  If a `getter` was input, this then causes errors with the `value` attribute.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

WilliamJamieson commented:
I think the issue here is that parameters should only be defined with both a getter and a setter. Where the getter and setter are inverses of each other. Otherwise, setting through the value property and then reading the value from the property would be different.

I feel the correct fix for this is for parameter to raise an error if only one of getter/setter is set during initialization of the parameter.

larrybradley commented:
@WilliamJamieson I thought so too, but there was some code and a test that allowed `setter` to be set without `getter`.  I've updated this PR to require both `getter` and `setter` (and updated the test and code to reflect that).

I also just noticed that if the getter should happen to return an array of shape (1,), `Parameter.value` will raise a `DeprecationWarning` for numpy 1.25+.  There no test for that yet, but I'll fix it in a separater PR.

larrybradley commented:
@WilliamJamieson I decided to just fix the numpy deprecation in this PR.  See the last 2 commits.
